metrics/README.md

```diff
@@ -0,0 +1,90 @@
+# Common metrics export interfaces for Knative
+
+_Note that this directory is currently in transition. See [the Plan](#the-plan)
+for details on where this is heading._
+
+## Current status
+
+The code currently uses OpenCensus to support exporting metrics to multiple
+backends. Currently, two backends are supported: Prometheus and Stackdriver.
+
+Metrics export is controlled by a ConfigMap called `config-observability` which
+is a key-value map with specific values supported for each of the Stackdriver
+and Prometheus backends. Hot-reload of the ConfigMap on a running process is
+supported by directly watching (via the Kubernetes API) the
+`config-observability` object. Configuration via environment is also supported
+for use by the `queue-proxy`, which runs with user permissions in the user's
+namespace.
+
+## Problems
+
+There are currently
+[6 supported Golang exporters for OpenCensus](https://opencensus.io/exporters/supported-exporters/go/).
+At least the Stackdriver exporter causes problems/failures if started without
+access to (Google) application defaut credentials. It's not clear that we want
+to build all of those backends into the core of `knative.dev/pkg` and all
+downstream dependents, and we'd like all the code shipped in `knative.dev/pkg`
+to be able to be tested without needing any special environment setup.
+
+With the current direct-integration setup, there needs to be initial and ongoing
+work in `pkg` (which should be high-value, low-churn code) to maintain and
+update stats exporters which need to be statically linked into ~all Knative
+binaries. This setup also causes problems for vendors who may want or need to
+perform an out-of-tree integration (e.g. proprietary or partially-proprietary
+monitoring stacks).
+
+Another problem is that each vendor's exporter requires different parameters,
+supplied as Golang `Options` methods which may require complex connections with
+the Knative ConfigMap. Two examples of this are secrets like API keys and the
+Prometheus monitoring port (which requires additional service/etc wiring).
+
+See also
+[this doc](https://docs.google.com/document/d/1t-aov3XrhobjCKW4kwScY44QAoahiwxoyXXFtZyL8jw/edit),
+where the plan was worked out.
+
+## The plan
+
+OpenCensus (and eventually OpenTelemetry) offers an sidecar or host-level agent
+with speaks the OpenCensus protocol and can proxy from this protocol to multiple
+backends.
+
+![OpenCensus Agent configuration](https://github.com/census-instrumentation/opencensus-service/raw/master/images/opencensus-service-deployment-models.png)
+
+**We will standardize on export to the OpenCensus agent, and encourage vendors
```

Comment:
I looked into the OpenCensus Agent vs OpenCensus Collector Service more. I believe we do **not** need the OpenCensus Agent as a sidecar as long as we instrument using the OpenCensus library.

I believe the agent's purpose is to allow code that is already implemented using Prometheus or Jaeger to use OpenCensus without modification.

Design doc for collector service: https://github.com/census-instrumentation/opencensus-service/blob/master/DESIGN.md#opencensus-collector

Reply:
So instead of using OpenCensus agent, we could do something like:
- Knative instruments using plain OpenCensus library, no exporters
- Knative always provides an OpenCensus Collector Service that supports Prometheus & Stackdriver out-of-the-box
- Some Knative controller converts 'config-observability' metrics fields into Collector service config?

Reply:
I think we can replace `config-observability` with a ConfigMap that configures the OpenCensus component.

I'd prefer to be able to use the default `localhost:55678` address rather than needing to configure a separate address for OpenCensus collection.

Reply:
Ugh, so after a bit of experimentation, it appears that standardizing on `localhost` is exactly what we don't want to do, because each netns (i.e. each Pod) has its own `localhost` and there's no good way to inject a port-forwarding rule into each one.

I'll update this section to highlight that the default `localhost:55678` **won't** work for common cases, and call out this is something to work on with OTel.